### DISCUSSION TITLE
Additional line separator

### DISCUSSION THREAD

**Discussion by nichtich** [pg-format/pg-formatter#22] [Ideas]
During writing test cases and cleaning up [the syntax diagram](https://github.com/gbv/pgraph/blob/main/docs/pg-grammar.md) I came across the need to quickly write a small graph in one line (implemented in the pgraph main branch). Adding an additional line separator is only one grammar rule. With this extension the graph (using implicit nodes for brevity)

~~~
a
b -> c
d -> e
~~~

could also be written

~~~
a ; b -> c ; d -> e
~~~

or less beautiful

~~~
a ;b -> c ;d -> e
~~~

I'm only unsure whether to also allow

~~~
a;b -> c;d -> e
~~~

because the latter would rewrite to disallow semicolon as last character of unquoted node identifier, label and/or property value.

Note that the following would not work anyway, as unquoted node identifiers are allowed to contain `->`:

~~~
a;b->c;d->e
~~~

**Comment by nichtich** [comment]
Meanwhile I've added `|` as line separator in [pgraphs](https://github.com/pg-format/pgraphs) and forbid this character in unquoted strings.

**Comment by hchiba1** [comment]
Interesting, but to be honest, I am not very willing to support it.
I have a feeling that the spec becomes complicated.

Is it possible to write PG in one line using `\n` (although it might be a bit ugly)?

**Reply by nichtich** [reply]
I've chosen `|` because it is not allowed in URIs. `\n` is only read as linebreak when part of a quoted string.

**Comment by hchiba1** [comment]
Is this notation indispensable?
My use cases will include `|` in values.
(This feature is a trade-off with unquoted characters.)

**Reply by nichtich** [reply]
How come `|` in unquoted values? I could also use another character but the choice is limited:

[RDF Turtle IRI syntax](https://www.w3.org/TR/rdf12-turtle/#terminals):

~~~
[30]	IRIREF	::=	'<' ([^#x00-#x20<>"{}|^`\] | UCHAR)* '>'
/* #x00=NULL #x01-#x1F=control codes #x20=space */
[43] UCHAR	::=	('\u' HEX HEX HEX HEX) | ('\U' HEX HEX HEX HEX HEX HEX HEX HEX)
[55]	HEX	::=	[0-9] | [A-F] | [a-f]
~~~

[SPARQL IRI syntax](https://www.w3.org/TR/sparql11-query/#QSynIRI)

> The set of RDF terms defined in RDF Concepts and Abstract Syntax includes RDF URI references while SPARQL terms include IRIs. RDF URI references containing "<", ">", '"' (double quote), space, "{", "}", "|", "\", "^", and "`" are not IRIs. 

~~~
[139]  	IRIREF	  ::=  	'<' ([^<>"{}|^`\]-[#x00-#x20])* '>'
~~~

I think `\` as line separator would work as well - it only has special meaning in quoted strings. Example:

~~~
a \ b \ c # three nodes
a\b\c # equivalent but less readable
a -> b :knows \ b -> c :knows # two edges
~~~

If there is a choice, I'd prefer `|` because `\` is also used for escape sequence in most other languages, so it will often need double escaping.

**Reply by hchiba1** [reply]
P.S. A simplified use case of `|`
```
b name:Bob|Robert
```
where `|` is not interpreted by PG parser, but maybe by applications.

**Reply by nichtich** [reply]
Looks like a short form of `b name:Bob name:Robert`.

**Comment by hchiba1** [comment]
If it is indispensable, I would like to push `;`.
Most people will guess what the `;` means.
(and will not find it strange to quote the strings containing `;`)

**Reply by hchiba1** [reply]
And I'm afraid a new variation of style can emerge (although it may not be a critical problem):

Someone may or may not end each statement with `;`
```
a : label
  key:val ;
a -> b
  key:val ;
b -> c
  key:val
```
I am not comfortable that `;` is sometimes added or omitted, but similar things happen in JavaScript as well!
In the end, it doesn’t look that strange, and it's acceptable to me.

**Reply by hchiba1** [reply]
Note that when `|` is chosen as a separator, the following should be valid:
```
a : label
  key:val |
a -> b
  key:val |
b -> c
  key:val |
```
Personally, I find it hard to readily accept.

**Reply by nichtich** [reply]
Sorry but I'd insist to use only the character `` <>"{}|^`\ `` for any any syntax extension. Additional characters such as `:` and `'` may be used when forbidden as first character of an unquoted string, but line separator is last character, so yet another special rule is needed. The last example you gave is valid in pgraph 0.5.2, parsing as

~~~
a :label key:val
a -> b key:val
b -> c key:val
~~~

The use case, however, is to allow for condense syntax:

~~~
a :label key:val | a -> b key:val | b -> c key:val
~~~

Talking about special rules: Replacing `,` with `|` for value lists would remove the need of a special rule to exclude `,` from unquoted values. Example together with line separator `\`:

~~~
a :label key:1|2 \ a -> b
~~~

But I fear it's too late to change list separator now.

**Comment by nichtich** [comment]
> Most people will guess what the `;` means.

Better keep syntax rules simple enough to fully explain on a page than expecting any previous assumptions. The *only* part we can assume is `->` to be read as an arrow, but that's it. Neither `--` nor `:` nor `"`or `'` are guessable because they each come with additional rules.

**Comment by nichtich** [comment]
It's better called *entity separator* or *statement separator*  because it can separate multiple nodes and/or edges on the same line. Depending on whether `|` or `\` is chosen as additional entity separator, all these are allowed:

~~~
a            # line break
b | a -> b   # entity separator
c |          # optional entity separator before line break
d|e          # condense
f            # end of input also terminates entity
~~~

~~~
a            # line break
b \ a -> b   # entity separator
c \          # optional entity separator before line break
d\e          # condense
f            # end of input also terminates entity
~~~

**Comment by nichtich** [comment]
I've removed statement separator from [specification draft](https://pg-format.github.io/specification/) to get to a stable 1.0.0. Characters `|` and `\` are reserved still so they can later be [used for extensions](https://pg-format.github.io/specification/#quoted-strings) such as statement separators, hyper-edges etc.
